tests/utf16.cpp

```diff
@@ -0,0 +1,55 @@
+// Copyright 2020 The Emscripten Authors.  All rights reserved.
+// Emscripten is available under two separate licenses, the MIT license and the
+// University of Illinois/NCSA Open Source License.  Both these licenses can be
+// found in the LICENSE file.
+
+#include <cassert>
+#include <emscripten.h>
+#include <stdio.h>
+#include <string>
+#include <vector>
+#include <wchar.h>
+
+static void testString(const char16_t* arg) {
+  // null-terminated string
+  std::u16string strz = arg;
+  char16_t* memory = new char16_t[strz.size() + 1]();
+
+  EM_ASM({
+      var strz = UTF16ToString($0);
+      out(strz);
+      var numBytesWritten = stringToUTF16(strz, $1, $2 * $3 + $3);
+      if (numBytesWritten != $2 * $3) throw 'stringToUTF16 wrote an invalid length ' + numBytesWritten;
+  }, strz.data(), memory, strz.size(), sizeof(char16_t));
+
+  // Compare memory to confirm that the string is intact after taking a route through JS side.
+  assert(std::equal(memory, memory + strz.size() + 1, strz.data()));
+
+  // non-null-terminated string
+  std::vector<char16_t> str(strz.begin(), strz.end());
```

Comment:
I wonder if it makes sense to make this vector a little longer and add some extra markers, so we can be sure its no null terminated?

Reply:
It would like to keep this test, because it triggers address sanitizer with the original code (which reads past the end of the vector). But I'll add an additional case with some garbage appended to the vector.